File: internal/daemon/testdata/without_comments/tmp.unmin.java
package workerpool;

import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;

import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.*;
import java.util.stream.*;

@java.lang.annotation.Retention(java.lang.annotation.RetentionPolicy.RUNTIME)
@java.lang.annotation.Target(java.lang.annotation.ElementType.METHOD)
@interface Tracked {
  String value() default "";
}

enum TaskPriority {
  LOW(0, "low"),
  MEDIUM(5, "medium"),
  HIGH(10, "high"),
  CRITICAL(20, "critical");
  private final int level;
  private final String label;

  TaskPriority(int level, String label) {
    this.level = level;
    this.label = label;
  }

  public int level() {
    return level;
  }

  public String label() {
    return label;
  }

  public String icon() {
    return switch (this) {
      case LOW -> ".";
      case MEDIUM -> "*";
      case HIGH -> "!";
      case CRITICAL -> "!!!";
    };
  }

  public String display() {
    return label + "(" + icon() + ")";
  }
}

record Task(long id, String name, TaskPriority priority, Map<String, Object> payload) {
  Task(long id, String name, TaskPriority priority) {
    this(id, name, priority, Map.of());
  }

  String describe() {
    var payloadInfo = payload.isEmpty() ? "none" : payload.keySet().stream().collect(joining(","));
    return "Task#%d[%s] priority=%s payload={%s}"
        .formatted(id, name, priority.display(), payloadInfo);
  }
}

record TaskResult(long taskId, boolean success, String output, Exception error) {
  static TaskResult ok(long taskId, String output) {
    return new TaskResult(taskId, true, output, null);
  }

  static TaskResult fail(long taskId, Exception error) {
    return new TaskResult(taskId, false, null, error);
  }
}

record PoolStats(int totalTasks, int completed, int failed, int activeWorkers) {}

sealed interface PoolEvent
    permits PoolEvent.TaskSubmitted, PoolEvent.TaskCompleted, PoolEvent.Shutdown {
  record TaskSubmitted(Task task) implements PoolEvent {}

  record TaskCompleted(TaskResult result) implements PoolEvent {}

  record Shutdown(String reason) implements PoolEvent {}
}

@FunctionalInterface
interface TaskHandler {
  TaskResult handle(Task task) throws Exception;
}

@FunctionalInterface
interface WorkerFactory {
  Worker create(int id, TaskHandler handler);
}

interface Loggable {
  String prefix();

  default void log(String message) {
    System.out.println("[" + prefix() + "] " + message);
  }

  default void error(String message, Throwable t) {
    System.err.println("[" + prefix() + "] ERROR: " + message);
    if (t != null) t.printStackTrace(System.err);
  }
}

class EventProcessor {
  static String describe(PoolEvent event) {
    if (event instanceof PoolEvent.TaskSubmitted ts) return "Submitted: " + ts.task().describe();
    else if (event instanceof PoolEvent.TaskCompleted tc && tc.result().success())
      return "OK: task " + tc.result().taskId();
    else if (event instanceof PoolEvent.TaskCompleted tc)
      return "FAIL: task " + tc.result().taskId();
    else if (event instanceof PoolEvent.Shutdown sd) return "Shutdown: " + sd.reason();
    return "Unknown event";
  }
}

class Worker implements Loggable {
  private final int id;
  private final TaskHandler handler;
  private volatile Task currentTask;

  Worker(int id, TaskHandler handler) {
    this.id = id;
    this.handler = handler;
  }

  @Override
  public String prefix() {
    return "Worker-" + id;
  }

  public int id() {
    return id;
  }

  @Tracked("execute")
  @SuppressWarnings("unchecked")
  public TaskResult execute(Task task) {
    currentTask = task;
    try {
      log("Processing: " + task.name());
      return handler.handle(task);
    } catch (Exception e) {
      error("Failed on task " + task.id(), e);
      return TaskResult.fail(task.id(), e);
    } finally {
      currentTask = null;
    }
  }

  public Optional<Task> currentTask() {
    return Optional.ofNullable(currentTask);
  }

  public String status() {
    return currentTask().map(t -> "busy with " + t.name()).orElse("idle");
  }
}

class WorkerPool implements Loggable, AutoCloseable {
  private final List<Worker> workers;
  private final ExecutorService executor;
  private final AtomicInteger completed = new AtomicInteger(0);
  private final AtomicInteger failed = new AtomicInteger(0);

  WorkerPool(int size, WorkerFactory factory, TaskHandler handler) {
    this.executor = Executors.newFixedThreadPool(size);
    var list = new ArrayList<Worker>(size);
    for (var i = 0; i < size; i++) {
      list.add(factory.create(i + 1, handler));
    }
    this.workers = Collections.unmodifiableList(list);
  }

  @Override
  public String prefix() {
    return "Pool";
  }

  public List<TaskResult> submitAll(List<? extends Task> tasks) {
    var sorted =
        tasks.stream()
            .sorted(
                Comparator.comparing(Task::priority, Comparator.comparingInt(TaskPriority::level))
                    .reversed()
                    .thenComparing(Task::name))
            .collect(toList());
    log("Submitting " + sorted.size() + " tasks to " + workers.size() + " workers");
    List<Future<TaskResult>> futures =
        sorted.stream()
            .map(
                task -> {
                  var worker = workers.get((int) (task.id() % workers.size()));
                  return executor.submit(
                      () -> {
                        var result = worker.execute(task);
                        if (result.success()) {
                          completed.incrementAndGet();
                        } else {
                          failed.incrementAndGet();
                        }
                        return result;
                      });
                })
            .collect(toList());
    return futures.stream()
        .map(
            f -> {
              try {
                return f.get(5, TimeUnit.SECONDS);
              } catch (Exception e) {
                return TaskResult.fail(
                    -1, e instanceof Exception ex ? ex : new RuntimeException(e));
              }
            })
        .toList();
  }

  public String summary(List<TaskResult> results) {
    var ok = results.stream().filter(TaskResult::success).count();
    var fail = results.stream().filter(r -> !r.success()).count();
    var outputs =
        results.stream().filter(TaskResult::success).map(TaskResult::output).collect(joining(", "));
    var totalLen =
        results.stream()
            .filter(TaskResult::success)
            .map(r -> r.output().length())
            .reduce(0, Integer::sum);
    return "Results: %d ok, %d failed, outputs=[%s], totalLen=%d"
        .formatted(ok, fail, outputs, totalLen);
  }

  public synchronized PoolStats stats() {
    return new PoolStats(
        completed.get() + failed.get(), completed.get(), failed.get(), workers.size());
  }

  public void drainTo(Collection<? super TaskResult> sink, List<TaskResult> results) {
    results.stream().filter(TaskResult::success).forEach(sink::add);
  }

  @Override
  public void close() {
    log("Shutting down pool");
    executor.shutdown();
    try {
      if (!executor.awaitTermination(10, TimeUnit.SECONDS)) executor.shutdownNow();
    } catch (InterruptedException e) {
      executor.shutdownNow();
      Thread.currentThread().interrupt();
    }
    var s = stats();
    log("Final: completed=%d, failed=%d".formatted(s.completed(), s.failed()));
  }
}

public class tmp {
  private static final String BANNER =
      """
      =============================
        Worker Pool Demo
        Java 17+ Features
      =============================\
      """;

  public static void main(String[] args) {
    System.out.println(BANNER);
    TaskHandler handler =
        task ->
            switch (task.priority()) {
              case CRITICAL -> {
                Thread.sleep(50);
                yield TaskResult.ok(task.id(), "URGENT:" + task.name().toUpperCase());
              }
              case HIGH, MEDIUM -> TaskResult.ok(task.id(), "processed:" + task.name());
              case LOW -> TaskResult.ok(task.id(), "deferred:" + task.name());
            };
    try (var pool = new WorkerPool(3, Worker::new, handler)) {
      var tasks =
          List.of(
              new Task(
                  1, "fetch-data", TaskPriority.HIGH, Map.of("url", "https://api.example.com")),
              new Task(2, "parse-json", TaskPriority.HIGH),
              new Task(3, "validate", TaskPriority.MEDIUM),
              new Task(4, "cleanup", TaskPriority.LOW, Map.of("path", "/tmp")),
              new Task(5, "alert", TaskPriority.CRITICAL));
      var results = pool.submitAll(tasks);
      results.stream()
          .filter(TaskResult::success)
          .map(r -> "  OK: task %d -> %s".formatted(r.taskId(), r.output()))
          .forEach(System.out::println);
      results.stream()
          .filter(r -> !r.success())
          .map(r -> "  FAIL: task %d -> %s".formatted(r.taskId(), r.error().getMessage()))
          .forEach(System.out::println);
      System.out.println(pool.summary(results));
      Optional<String> firstOutput =
          results.stream().filter(TaskResult::success).map(TaskResult::output).findFirst();
      System.out.println("First output: " + firstOutput.orElse("none"));
      PoolEvent event = new PoolEvent.TaskCompleted(results.getFirst());
      System.out.println("Event: " + EventProcessor.describe(event));
      var sink = new ArrayList<Object>();
      pool.drainTo(sink, results);
      System.out.println("Drained " + sink.size() + " results");
      var allOk = results.stream().allMatch(TaskResult::success);
      System.out.println("Status: " + (allOk ? "ALL PASSED" : "SOME FAILED"));
    }
  }
}

File: internal/daemon/testdata/without_comments/tmp.unmin.kt
package workerpool

import java.util.concurrent.atomic.AtomicInteger
import kotlinx.coroutines.*

enum class TaskPriority(val level: Int, val label: String) {
  LOW(0, "low") {
    override fun icon() = "."
  },
  MEDIUM(5, "medium") {
    override fun icon() = "*"
  },
  HIGH(10, "high") {
    override fun icon() = "!"
  },
  CRITICAL(20, "critical") {
    override fun icon() = "!!!"
  };

  abstract fun icon(): String

  fun display() = label + "(" + icon() + ")"
}

typealias TaskId = Long

typealias TaskHandler = suspend (Task) -> TaskResult

typealias WorkerFactory = (Int) -> Worker

sealed interface TaskResult {
  val taskId: TaskId

  data class Success(override val taskId: TaskId, val output: String) : TaskResult

  data class Failure(override val taskId: TaskId, val error: Throwable) : TaskResult

  data class Retry(override val taskId: TaskId, val attempt: Int, val maxAttempts: Int) : TaskResult
}

data class Task(
    val id: TaskId,
    val name: String,
    val priority: TaskPriority,
    val payload: Map<String, Any?> = emptyMap(),
    val timeoutMs: Long = 5000L,
)

data class PoolStats(
    val totalTasks: Int,
    val completed: Int,
    val failed: Int,
    val activeWorkers: Int,
)

sealed class PoolEvent {
  data class TaskSubmitted(val task: Task) : PoolEvent()

  data class TaskCompleted(val result: TaskResult) : PoolEvent()

  data class WorkerStateChange(val workerId: Int, val running: Boolean) : PoolEvent()

  object PoolShutdown : PoolEvent()
}

object WorkerIdGenerator {
  private val counter = AtomicInteger(0)

  fun next(): Int = counter.incrementAndGet()

  fun reset() {
    counter.set(0)
  }
}

interface Logger {
  fun log(message: String)

  fun error(message: String, throwable: Throwable? = null)
}

class ConsoleLogger(private val prefix: String) : Logger {
  override fun log(message: String) = println("[" + prefix + "] " + message)

  override fun error(message: String, throwable: Throwable?) {
    System.err.println("[" + prefix + "] ERROR: " + message)
    throwable?.printStackTrace(System.err)
  }
}

class Worker(
    val id: Int,
    private val handler: TaskHandler,
    loggerImpl: Logger = ConsoleLogger("Worker-" + id),
) : Logger by loggerImpl {
  val displayName: String by lazy { "Worker-#" + id + "[" + Thread.currentThread().name + "]" }

  private var currentTask: Task? = null

  suspend fun execute(task: Task): TaskResult {
    currentTask = task
    return try {
      val name = currentTask?.name ?: "unknown"
      log(displayName + " processing: " + name)
      currentTask?.let { t ->
        t.also { log("Task " + it.id + " payload size: " + it.payload.size) }
        handler(t)
      } ?: TaskResult.Failure(task.id, IllegalStateException("Task became null"))
    } catch (e: Exception) {
      val msg = runCatching { e.message!! }.getOrElse { "No message" }
      error("Failed: " + msg, e)
      TaskResult.Failure(task.id, e)
    } finally {
      currentTask = null
    }
  }

  fun status(): String = currentTask.run { this?.let { "busy with " + it.name } ?: "idle" }
}

fun buildTaskList(block: MutableList<Task>.() -> Unit): List<Task> {
  return mutableListOf<Task>().apply(block)
}

fun List<Task>.sortedByPriority(): List<Task> = sortedByDescending { it.priority.level }

fun Task.describe(): String =
    "Task(id=" +
        id +
        ", name='" +
        name +
        "', priority=" +
        priority.display() +
        ", timeout=" +
        timeoutMs +
        "ms)"

fun TaskResult.isTerminal(): Boolean =
    when (this) {
      is TaskResult.Success -> true
      is TaskResult.Failure -> true
      is TaskResult.Retry -> attempt >= maxAttempts
    }

inline fun <reified T : TaskResult> List<TaskResult>.filterResults(): List<T> =
    filterIsInstance<T>()

inline fun <reified T : Any> Task.payloadValue(key: String): T? = payload[key] as? T

fun describeResult(result: TaskResult): String =
    when (result) {
      is TaskResult.Success -> "OK: " + result.output
      is TaskResult.Failure ->
          when {
            result.error is IllegalArgumentException -> "Bad input: " + result.error.message
            else -> "Error: " + result.error
          }
      is TaskResult.Retry ->
          when {
            result.attempt >= result.maxAttempts ->
                "Giving up after " + result.attempt + " attempts"
            else -> "Retry " + result.attempt + "/" + result.maxAttempts
          }
    }

fun categorizePriority(p: TaskPriority): String =
    when (p) {
      TaskPriority.LOW,
      TaskPriority.MEDIUM -> "normal"
      TaskPriority.HIGH -> "elevated"
      TaskPriority.CRITICAL -> "urgent"
    }

fun createTasks(vararg names: String, priority: TaskPriority = TaskPriority.MEDIUM): List<Task> =
    names.mapIndexed { idx, name ->
      Task(id = idx.toLong(), name = name, priority = priority)
    }

fun workerIds(poolSize: Int): List<Int> = (1..poolSize).toList()

fun batchRanges(total: Int, batchSize: Int): List<IntRange> =
    (0 until total step batchSize).map { it until minOf(it + batchSize, total) }

class WorkerPool(
    private val size: Int,
    private val factory: WorkerFactory,
    loggerImpl: Logger = ConsoleLogger("Pool"),
) : Logger by loggerImpl {
  private val workers: List<Worker> by lazy { workerIds(size).map { factory(it) } }

  private val completedCount = AtomicInteger(0)

  private val failedCount = AtomicInteger(0)

  companion object {
    const val DEFAULT_SIZE = 4

    const val MAX_SIZE = 64

    fun default(handler: TaskHandler): WorkerPool =
        WorkerPool(DEFAULT_SIZE, { id -> Worker(id, handler) })
  };

  fun stats(): PoolStats {
    val (total, completed, failed) =
        Triple(completedCount.get() + failedCount.get(), completedCount.get(), failedCount.get())
    return PoolStats(total, completed, failed, workers.size)
  }

  suspend fun submitAll(tasks: List<Task>): List<TaskResult> = coroutineScope {
    val sorted = tasks.sortedByPriority()
    log("Submitting " + sorted.size + " tasks to " + size + " workers")
    sorted
        .mapIndexed { index, task ->
          val worker = workers[index % workers.size]
          async {
            log("Dispatching '" + task.name + "' to worker " + worker.id)
            val result = worker.execute(task)
            when (result) {
              is TaskResult.Success -> completedCount.incrementAndGet()
              is TaskResult.Failure -> failedCount.incrementAndGet()
              is TaskResult.Retry -> {}
            }
            result
          }
        }
        .awaitAll()
  }

  fun submitBlocking(tasks: List<Task>): List<TaskResult> = runBlocking { submitAll(tasks) }

  fun shutdown() {
    log("Shutting down pool with " + workers.size + " workers")
    val (_, completed, failed, active) = stats()
    log("Final stats: completed=" + completed + ", failed=" + failed + ", active=" + active)
  }
}

class PoolConfig {
  var poolSize: Int = WorkerPool.DEFAULT_SIZE

  var logPrefix: String = "Pool"

  var handler: TaskHandler = { TaskResult.Success(it.id, "done") }

  fun handler(block: TaskHandler) {
    this.handler = block
  }
}

fun workerPool(configure: PoolConfig.() -> Unit): WorkerPool {
  val config = PoolConfig().apply(configure)
  return WorkerPool(
      size = config.poolSize,
      factory = { id -> Worker(id, config.handler) },
      loggerImpl = ConsoleLogger(config.logPrefix),
  )
}

fun main() {
  val pool = workerPool {
    poolSize = 3
    logPrefix = "Demo"
    handler { task ->
      delay(100)
      val key = task.payloadValue<String>("key")
      val output = key?.uppercase() ?: ("processed-" + task.name)
      TaskResult.Success(task.id, output)
    }
  }
  val tasks = buildTaskList {
    addAll(createTasks("fetch-data", "parse-json", "validate", priority = TaskPriority.HIGH))
    add(Task(100, "cleanup", TaskPriority.LOW, mapOf("key" to "temp")))
    add(Task(101, "report", TaskPriority.CRITICAL, timeoutMs = 10_000L))
  }
  for ((id, name, priority, payload, timeout) in tasks) {
    println(
        "Queued: #" +
            id +
            " '" +
            name +
            "' [" +
            priority.display() +
            "] timeout=" +
            timeout +
            "ms payload=" +
            payload
    )
  }
  val batches = batchRanges(tasks.size, 2)
  println("Processing in " + batches.size + " batches: " + batches)
  val results = pool.submitBlocking(tasks)
  val successes = results.filterResults<TaskResult.Success>()
  val failures = results.filterResults<TaskResult.Failure>()
  successes.forEach { println("  OK: " + describeResult(it)) }
  failures.forEach { println("  FAIL: " + describeResult(it)) }
  val firstSuccess = successes.firstOrNull()
  val outputUpper = firstSuccess?.output?.uppercase() ?: "none"
  println("First success output: " + outputUpper)
  val parsed: Int =
      try {
        "notAnInt".toInt()
      } catch (_: NumberFormatException) {
        -1
      }
  println("Parsed fallback: " + parsed)
  pool.shutdown()
}
